core/msg.c

```diff
@@ -56,7 +56,9 @@ static int queue_msg(thread_t *target, const msg_t *m)
     *dest = *m;
 #if MODULE_CORE_THREAD_FLAGS
     target->flags |= THREAD_FLAG_MSG_WAITING;
-    thread_flags_wake(target);
+    if (thread_flags_wake(target)) {
+        thread_yield_higher();
```

Comment:
Should the same also be done in [line 155](https://github.com/RIOT-OS/RIOT/blob/master/core/msg.c#L153)?

And if the answer to this is yes, why not move the `thread_yield_higher()` to `thread_flags_wake()` since those two are the only locations where `thread_flags_wake()` is called.

Reply:
in line 155 there's *always* a thread_yield following, below the `irq_restore()`.
The yield is not in thread_flags wake because when it was implemented, thread_yield was not isr safe, not sure it is now.
Also, there's use-cases where many threads need to be woken up. maybe this needs to be revisited.

the thread_yield_higher() is frustrating, I think we need to go through all the platforms and split those. IIRC most of them do the equivalent of "if irq_is_in() sched_context_switch_request=1 else platform_stuff()", that can maybe be pulled out.